### PR TITLE
Add IndexMapping Tool

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Adds an IndexMappingTool.   See previous dev branch PR: #1609.  

It looks like it's part of pending PR #1812 but needed to unblock end-to-end testing.
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by ylwu-amzn — APPROVED** [review]

**Review by arjunkumargiri — COMMENTED** [review]
Can we do manual integration test by adding index mapping tool to chat agent and validate that the tool is able to integrate with LLM as expected?

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+                List<ModelTensors> mlModelOutputs = (List<ModelTensors>) o;
+                return mlModelOutputs.get(0).getMlModelTensors().get(0).getDataAsMap().get("response");
+            }
+        };
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        @SuppressWarnings("unchecked")
+        List<String> indexList = parameters.containsKey("index")
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java:69]: This is how the index mapping API works: https://opensearch.org/docs/latest/api-reference/index-apis/put-mapping/ 
The tool uses the feature to expand wildcards:
> Expands wildcard expressions to different indexes. Combine multiple values with commas. Available values are all (match all indexes), open (match open indexes), closed (match closed indexes), hidden (match hidden indexes), and none (do not accept wildcard expressions), which must be used with open, closed, or both. Default is open.

So a single index passed as a string works.   As does a comma delimited list.  As does a wildcard.  As does the word "all".

**Comment by dbwiddis** [comment]
> Can we do manual integration test by adding index mapping tool to chat agent and validate that the tool is able to integrate with LLM as expected?

I'm already doing this with the testing framwork currently at https://github.com/joshuali925/observability-langchain but soon to be moved to this org.

Sample output, including a hallucination:

```
Running test: 9E036FA47
Received: After reviewing the available information, the data fields available for the aircraft-match index are id, aircraft_id, and match_id.
Expected: The data fields available for the aircraft-match index are id, aircraft_id, and match_id
Score: 1

Running test: AC1D6C46C
Received: After reviewing the available information, the data type of the '%_Change_2007' field in the 'aircraft-airport' index is a string.
Expected: The data type of the '%_Change_2007' field in the 'aircraft-airport' index is a string.
Score: 1

Running test: CBBC09702
Received: After reviewing the information gathered, the fields for the aircraft-aircraft index are name, model, and year.
Expected: The fields for the aircraft-aircraft index are Aircraft, Aircraft_ID, Description, Max_Gross_weight, Max_disk_Loading, and Total_disk_area
Score: 0
```

**Comment by dbwiddis** [comment]
- JDK 21 test failures are as a result of https://github.com/opensearch-project/flow-framework/issues/426
- Mend vulnerability is not part of this PR

**Review by arjunkumargiri — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+@ToolAnnotation(IndexMappingTool.TYPE)
+public class IndexMappingTool implements Tool {
+    public static final String TYPE = "IndexMappingTool";
+    private static final String DEFAULT_DESCRIPTION = String
+        .join(
+            " ",
```
Comment (lezzago) [inline ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java:38]: Why do we need an empty space?

**Review by lezzago — APPROVED** [review]

File: ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java (hunk 2 above)
Comment (dbwiddis) [reply ml-algorithms/src/main/java/org/opensearch/ml/engine/tools/IndexMappingTool.java:38]: This is the delimiter in string join, adding a space between the sentences.  I thought this syntax was easier to read than a bunch of concatenated strings.  It's a one-time static constant so performance isn't really an issue.

I could append the strings with a + and try to remember to add a space at the end or beginning of each subsequent one....

**Review by rbhavna — APPROVED** [review]

**Review by ylwu-amzn — APPROVED** [review]

**Pull request merged**